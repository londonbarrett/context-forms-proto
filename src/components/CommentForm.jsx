import * as React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { Form, validators } from './ContextForms';
import TextfieldInput from './inputs/TextfieldInput';
import CheckboxInput from './inputs/CheckboxInput';
import FieldError from './FieldError';
import RatingInput from './inputs/RatingInput';
import SelectInput from './inputs/SelectInput';
import Reset from './Reset';
import Submit from './Submit';
import FormErrors from './FormErrors';

const StyledForm = styled(Form)`
  margin: 4rem auto;
  width: 400px;
`;

const Field = styled.div`
  margin: 1rem 0;
`;

const Label = styled.label`
  color: #007367;
  display: block;
  font-size: 1.3rem;
  margin: 1rem 0;
  text-align: left;
`;

const Actions = Field.extend`
  margin-top: 2rem;
  text-align: right;
`;

class CommentForm extends React.PureComponent {
  handleRegionChange = (event) => {
    console.log('COMMENT_FORM', event.target.value);
  }

  render() {
    const {
      firstName,
      lastName,
      age,
      accept,
      rating,
      onSubmit,
    } = this.props;
    return (
      <StyledForm
        onSubmit={onSubmit}
      >
        <Field>
          <Label htmlFor="firstName">
            First Name
          </Label>
          <TextfieldInput
            id="firstName"
            name="firstName"
            validators={[
              validators.isNotEmpty('First Name should not be empty'),
            ]}
            value={firstName}
          />
          <FieldError name="firstName" />
        </Field>
        <Field>
          <Label htmlFor="lastName">
            Last Name
          </Label>
          <TextfieldInput
            id="lastName"
            name="lastName"
            value={lastName}
          />
        </Field>
        <Field>
          <Label htmlFor="age">
            Age
          </Label>
          <TextfieldInput
            id="age"
            name="age"
            validators={[
              validators.isNotEmpty('Age should not be empty'),
              validators.isNumber('Age should be a number'),
              validators.greaterThan(150)('Age should be under 150'),
            ]}
            value={age}
          />
          <FieldError name="age" />
        </Field>
        <Field>
          <Label htmlFor="accept">
            Terms & Conditions
          </Label>
          <CheckboxInput
            id="accept"
            label="Do you accept terms and contidions"
            name="accept"
            validators={[
              validators.isRequired('Terms & Conditions must be accepted'),
            ]}
            value={accept}
          />
          <FieldError name="accept" />
        </Field>
        <Field>
          <Label htmlFor="region">
            Region
          </Label>
          <SelectInput
            name="Region"
            options={[
              {
                label: 'Africa',
                value: 1,
              },
              {
                label: 'America',
                value: 2,
              },
              {
                label: 'Europe',
                value: 3,
              },
            ]}
            onChange={this.handleRegionChange}
          />
        </Field>
        <Field>
          <Label htmlFor="rating">
            Rating
          </Label>
          <RatingInput
            id="rating"
            name="rating"
            validators={[
              validators.isRequired('Rating field is required'),
            ]}
            value={rating}
          />
        </Field>
        <FormErrors />
        <Actions>
          <Reset />
          <Submit value="Send" />
        </Actions>
      </StyledForm>
    );
  }
}

CommentForm.propTypes = {
  firstName: PropTypes.string,
  lastName: PropTypes.string,
  age: PropTypes.string,
  accept: PropTypes.bool,
  rating: PropTypes.number,
  onSubmit: PropTypes.func,
};

CommentForm.defaultProps = {
  firstName: undefined,
  lastName: undefined,
  age: undefined,
  accept: undefined,
  rating: undefined,
  onSubmit: undefined,
};

export default CommentForm;
