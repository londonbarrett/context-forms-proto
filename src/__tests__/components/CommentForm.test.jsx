import React from 'react';
import { shallow } from 'enzyme';
import { render, fireEvent } from 'react-testing-library';
import CommentForm from '../../components/CommentForm';

it('Renders correctly', () => {
  const wrapper = shallow(<CommentForm />);
  expect(wrapper).toMatchSnapshot();
});

it('Renders all fields', () => {
  const wrapper = shallow(<CommentForm />);
  const fields = wrapper.find('FormInput').length;
  expect(fields).toEqual(5);
});

it('Submits form, and receive values', () => {
  const handleOnSubmit = jest.fn();
  const { getByText } = render(
    <CommentForm
      firstName="Joe"
      lastName="Strummer"
      age="66"
      accept
      rating={5}
      onSubmit={handleOnSubmit}
    />,
  );
  const submit = getByText(
    (content, element) => element.getAttribute('type') === 'submit',
  );
  fireEvent(submit, new MouseEvent('click'));
  expect(handleOnSubmit).toBeCalledWith({
    firstName: 'Joe',
    lastName: 'Strummer',
    age: '66',
    accept: true,
    rating: 5,
  });
});

it('Doesnt submit form', () => {
  const handleOnSubmit = jest.fn();
  const { getByText } = render(
    <CommentForm
      firstName="Joe"
      lastName="Strummer"
      age="666"
      accept
      rating={5}
      onSubmit={handleOnSubmit}
    />,
  );
  const submit = getByText(
    (content, element) => element.getAttribute('type') === 'submit',
  );
  fireEvent(submit, new MouseEvent('click'));
  expect(handleOnSubmit).not.toBeCalled();
});
