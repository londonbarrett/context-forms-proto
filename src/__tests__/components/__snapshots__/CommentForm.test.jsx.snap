// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders correctly 1`] = `
<Styled(Form)>
  <styled.div>
    <styled.label
      htmlFor="firstName"
    >
      First Name
    </styled.label>
    <FormInput
      id="firstName"
      name="firstName"
      validators={
        Array [
          [Function],
        ]
      }
    />
    <Wrapper
      name="firstName"
    />
  </styled.div>
  <styled.div>
    <styled.label
      htmlFor="lastName"
    >
      Last Name
    </styled.label>
    <FormInput
      id="lastName"
      name="lastName"
    />
  </styled.div>
  <styled.div>
    <styled.label
      htmlFor="age"
    >
      Age
    </styled.label>
    <FormInput
      id="age"
      name="age"
      validators={
        Array [
          [Function],
          [Function],
          [Function],
        ]
      }
    />
    <Wrapper
      name="age"
    />
  </styled.div>
  <styled.div>
    <styled.label
      htmlFor="accept"
    >
      Terms & Conditions
    </styled.label>
    <FormInput
      id="accept"
      label="Do you accept terms and contidions"
      name="accept"
      validators={
        Array [
          [Function],
        ]
      }
    />
    <Wrapper
      name="accept"
    />
  </styled.div>
  <styled.div>
    <styled.label
      htmlFor="rating"
    >
      Rating
    </styled.label>
    <FormInput
      id="rating"
      name="rating"
      validators={
        Array [
          [Function],
        ]
      }
    />
  </styled.div>
  <Wrapper />
  <styled.div>
    <Reset />
    <Wrapper
      value="Send"
    />
  </styled.div>
</Styled(Form)>
`;
